lxdock/provisioners/shell.py

```diff
@@ -15,25 +15,29 @@ class ShellProvisioner(Provisioner):
         'side': Any('guest', 'host'),
     }
 
-    def provision(self):
+    def provision_single(self, guest):
         """ Executes the shell commands in the guest container or in the host. """
         if 'script' in self.options and self._is_for_guest:
             # First case: we have to run the script inside the container. So the first step is
             # to copy the content of the script to a temporary file in the container, ensure
             # that the script is executable and then run the script.
             guest_scriptpath = os.path.join('/tmp/', os.path.basename(self.options['script']))
             with open(self.homedir_expanded_path(self.options['script'])) as fd:
-                self.guest.lxd_container.files.put(guest_scriptpath, fd.read())
-            self.guest.run(['chmod', '+x', guest_scriptpath])
-            self.guest.run([guest_scriptpath, ])
+                guest.lxd_container.files.put(guest_scriptpath, fd.read())
+            guest.run(['chmod', '+x', guest_scriptpath])
+            guest.run([guest_scriptpath, ])
         elif 'script' in self.options and self._is_for_host:
             # Second case: the script is executed on the host side.
             self.host.run([self.homedir_expanded_path(self.options['script']), ])
         elif 'inline' in self.options:
             # Final case: we run a command directly inside the container or outside.
-            host_or_guest = getattr(self, self._side)
+            host_or_guest = self.host if self._side == 'host' else guest
             host_or_guest.run(['sh', '-c', self.options['inline']])
 
+    def setup(self):
+        # nothing to set up, avoid spurious messages with this override.
```

Comment:
Three options I see:

1. Removing the logging message at the beginning of `setup_single()`, which allows us to remove the empty override without side effect.
2. Add a `has_setup` class member to `Provisioner` and set it to true in puppet and ansible.
3. Simply remove the override and accept that our shell provisoner will spew a useless "Preparing {0} up for provisioner {1}" message. That's the only effect the removal will have.

All in all, the `has_setup` thing wouldn't be so much more elegant than the current situation, considering that the shell provisioner is likely the only provisioner we'd have that requires no setup.